### PR TITLE
Alpha Autonomous mode, waypoint following, heading hold

### PR DESCRIPTION
**Pull request by ArthurAllshire**

### CODE REVIEW

File: robot.py [hunk 1]
```diff
@@ -57,18 +67,37 @@ def createObjects(self):
         # boilerplate setup for the joystick
         self.joystick = wpilib.Joystick(0)
 
+        self.cube_switch = wpilib.DigitalInput(0)
+
         self.spin_rate = 5
 
+        NetworkTables.initialize()
```
Comment (auscompgeek) [inline robot.py]: wtf is this

File: pyswervedrive/swervechassis.py [hunk 2]
```diff
@@ -97,21 +146,37 @@ def execute(self):
         self.odometry_x_vel = v_x
         self.odometry_y_vel = v_y
         self.odometry_z_vel = v_z
+        print("odometry x %s, y %s" % (self.odometry_x, self.odometry_y))
 
     def robot_movement_from_odometry(self, odometry_outputs):
```
Comment (danieloconell) [inline pyswervedrive/swervechassis.py]: Where is the docstring.

File: automations/motion.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                [x_in_meters, y_in_meters, orientation_in_radians, speed_in_meters]
+        """
+        self.waypoints = waypoints
+        self.pursuit.set_waypoints(waypoints)
+        self.enabled = True
+        self.chassis.heading_hold_on()
+        self.update_heading_profile()
+        self.heading_error_i = 0
+
+    def update_heading_profile(self):
```
Comment (danieloconell) [inline automations/motion.py]: Doc string

File: automations/motion.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            vy = speed_sp * math.sin(direction_of_motion)
+
+            # self.chassis.set_velocity_heading(vx, vy, self.waypoints[self.waypoint_idx+1][2])
+            self.chassis.set_inputs(vx, vy, heading_output)
+
+            if over:
+                self.enabled = False
+
+    @property
+    def waypoint_idx(self):
```
Comment (danieloconell) [inline automations/motion.py]: Docstring

File: autonomous/autonomous.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+            self.motion.set_waypoints([[self.chassis.odometry_x, self.chassis.odometry_y, angle, 0],
+                                       [2.5, 0, math.pi/2, 1.5],
+                                       [2.5, 1, math.pi/2, 1.5]])
+        if not self.motion.enabled:
+            print("going to 'intake cube'")
+            # self.next_state("search_for_cube")
+            self.next_state_now("intake_cube")
+
+    @state
+    def intake_cube(self):
```
Comment (danieloconell) [inline autonomous/autonomous.py]: Docstring

File: autonomous/autonomous.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+            self.next_state("search_for_cube")
+
+
+class LeftSwitchAndScale(SwitchAndScale):
+    """The switch and scale strategy when we start on the left"""
+    MODE_NAME: 'Switch and scale - left start'
+    start_position = 3  # change to left start position
+
+    @state(first=True)
+    def go_to_scale(self):
```
Comment (danieloconell) [inline autonomous/autonomous.py]: Docstring

File: autonomous/autonomous.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+                    [6, 2.1, 4.71239, 1],
+                    [4.2, 2.1, 4.71239, 1]],  # CrossSwitchSameScale
+                   [[5, -3, 0, 1],
+                    [6, 2, 0, 1],
+                    [7.5, 1.8, 0, 1],
+                    [5.5, 2, 3.92699, 1],
+                    [5.25, 1.5, 3.14159, 1]]]  # CrossSwitchCrossScale
+
+    @state(first=True)
+    def go_to_scale(self):
```
Comment (danieloconell) [inline autonomous/autonomous.py]: Docstring

File: autonomous/autonomous.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+                    [5.25, 1.80, 2.35619, 1],
+                    [7.50, 2, 0, 1]],  # coordinates for same scale
+                   [[5, 3, 0, 1],
+                    [6, -2, 0, 1],
+                    [7.5, -1.8, 0, 1],
+                    [5.5, -2, 3.92699, 1],
+                    [7.5, -2, 0, 1]]]  # coordinates for cross scale
+
+    @state(first=True)
+    def go_to_scale(self):
```
Comment (danieloconell) [inline autonomous/autonomous.py]: Docstring

File: autonomous/autonomous.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+                    [5.25, -1.80, 2.35619, 1],
+                    [7.50, -2, 0, 1]],  # coordinates for same scale
+                   [[5, -3, 0, 1],
+                    [6, 2, 0, 1],
+                    [7.5, 1.8, 0, 1],
+                    [5.5, 2, 3.92699, 1],
+                    [7.5, 2, 0, 1]]]  # coordinates for cross scale
+
+    @state(first=True)
+    def go_to_scale(self):
```
Comment (danieloconell) [inline autonomous/autonomous.py]: Docstring

File: components/vision.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    def execute(self):
+        """Run at the end of every control loop iteration."""
+        pass
+
+    def new_value(self, entry, key, value, param):
+        start_time = time.time()
+        if key == "/vision/info":
+            self.unpack_info(value, start_time)
+
+    def format_list(self, lst, n):
```
Comment (danieloconell) [inline components/vision.py]: Docstring

File: components/vision.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+    def on_disable(self):
+        """This is called whenever the robot transitions to disabled mode."""
+        pass
+
+    def execute(self):
+        """Run at the end of every control loop iteration."""
+        pass
+
+    def new_value(self, entry, key, value, param):
```
Comment (danieloconell) [inline components/vision.py]: Docstring

File: components/vision.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            self.unpack_info(value, start_time)
+
+    def format_list(self, lst, n):
+        out = [lst[i:i+n] for i in range(0, len(lst), n)]
+        t = out[-1][0]
+        out.pop(-1)
+        out.append(t)
+        return out
+
+    def unpack_info(self, info, start_time):
```
Comment (danieloconell) [inline components/vision.py]: Docstring

File: components/vision.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            return
+
+        info = np.array(info)
+        info = self.format_list(info, 3)
+        total_time = (time.time() - start_time) + float(info[-1])
+        info[-1] = total_time
+
+        self.vision = info
+
+    def largest_cube(self):
```
Comment (danieloconell) [inline components/vision.py]: Docstring

File: pyswervedrive/swervechassis.py [hunk 14]
```diff
@@ -18,23 +18,62 @@ def __init__(self):
...
         self.modules = [self.module_a, self.module_b, self.module_c, self.module_d]
 
+        self.odometry_x = 0
+        self.odometry_y = 0
+        self.odometry_theta = 0
+        self.odometry_x_vel = 0
+        self.odometry_y_vel = 0
+        self.odometry_z_vel = 0
+
+    def set_heading_sp_current(self):
```
Comment (danieloconell) [inline pyswervedrive/swervechassis.py]: Docstring

File: pyswervedrive/swervechassis.py [hunk 15]
```diff
@@ -18,23 +18,62 @@ def __init__(self):
...
+        self.set_heading_sp(self.bno055.getAngle())
+
+    def set_heading_sp(self, setpoint):
```
Comment (danieloconell) [inline pyswervedrive/swervechassis.py]: Docstring

File: pyswervedrive/swervechassis.py [hunk 16]
```diff
@@ -18,23 +18,62 @@ def __init__(self):
...
+        self.heading_pid.setSetpoint(setpoint)
+        self.heading_pid.enable()
+        self.momentum = False
+
+    def heading_hold_on(self):
```
Comment (danieloconell) [inline pyswervedrive/swervechassis.py]: Docstring

File: pyswervedrive/swervechassis.py [hunk 17]
```diff
@@ -18,23 +18,62 @@ def __init__(self):
...
+        self.set_heading_sp_current()
+        self.heading_pid.reset()
+        self.hold_heading = True
+
+    def heading_hold_off(self):
```
Comment (danieloconell) [inline pyswervedrive/swervechassis.py]: Docstring

File: pyswervedrive/swervechassis.py [hunk 18]
```diff
@@ -124,3 +189,8 @@ def field_orient(vx, vy, heading):
         oriented_vx = vx * math.cos(heading) + vy * math.sin(heading)
         oriented_vy = -vx * math.sin(heading) + vy * math.cos(heading)
         return oriented_vx, oriented_vy
+
+
+class ChassisPIDOutput(PIDOutput):
+    def pidWrite(self, output):
```
Comment (danieloconell) [inline pyswervedrive/swervechassis.py]: Docstring

File: pyswervedrive/swervemodule.py [hunk 19]
```diff
@@ -192,16 +198,19 @@ def set_velocity(self, vx, vy):
...
+                                     -velocity*self.drive_velocity_to_native_units)
             else:
                 self.drive_motor.set(ctre.ControlMode.Velocity, 0)
         else:
-            self.drive_motor.set(ctre.ControlMode.Velocity, velocity*self.drive_velocity_to_native_units)
+            self.drive_motor.set(ctre.ControlMode.Velocity,
+                                 velocity*self.drive_velocity_to_native_units)
 
     @property
     def current_azimuth(self):
```
Comment (danieloconell) [inline pyswervedrive/swervemodule.py]: Docstring

File: robot.py [hunk 20]
```diff
@@ -3,23 +3,33 @@
...
+from components.vision import Vision
 from pyswervedrive.swervechassis import SwerveChassis
 from pyswervedrive.swervemodule import SwerveModule
 from utilities.bno055 import BNO055
 from utilities.functions import rescale_js
 
+import math
+
 
 class Robot(magicbot.MagicRobot):
```
Comment (danieloconell) [inline robot.py]: Docsstring not comments

File: tests/test_vector_pursuit.py [hunk 21]
```diff
@@ -0,0 +1,90 @@
+from utilities.vector_pursuit import VectorPursuit
+import numpy as np
+import math
+from magicbot import MagicRobot
+
+
+def position_delta_x_y(theta, speed, robot_x, robot_y, orientation):
```
Comment (danieloconell) [inline tests/test_vector_pursuit.py]: Docstring

File: tests/test_vector_pursuit.py [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+    vy = speed * math.sin(theta)
+    oriented_vx = vx * math.cos(orientation) + vy * math.sin(orientation)
+    oriented_vy = -vx * math.sin(orientation) + vy * math.cos(orientation)
+
+    robot_x += MagicRobot.control_loop_wait_time*oriented_vx
+    robot_y += MagicRobot.control_loop_wait_time*oriented_vy
+    return robot_x, robot_y
+
+
+def test_single_waypoint_converge():
```
Comment (danieloconell) [inline tests/test_vector_pursuit.py]: Docstring

File: tests/test_vector_pursuit.py [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+                                              robot_x, robot_y, orientation)
+        if over:
+            break
+        last_speed = speed
+
+    assert robot_x > 2
+    assert abs(robot_y) < 0.1
+
+
+def test_multi_waypoint_converge():
```
Comment (danieloconell) [inline tests/test_vector_pursuit.py]: Docstring

File: tests/test_vector_pursuit.py [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+                                              robot_x, robot_y, orientation)
+        if over:
+            break
+        last_speed = speed
+
+    assert abs(robot_x-1) < 0.1
+    assert robot_y > 2
+
+
+def test_speed_control():
```
Comment (danieloconell) [inline tests/test_vector_pursuit.py]: Docstring

File: utilities/profile_generator.py [hunk 25]
```diff
@@ -0,0 +1,217 @@
...
+def sign(x):
+    if x >= 0:
+        return 1
+    else:
+        return -1
+
+
+def cubic_generator(keypoints):
+    """Return a function that returns the distance and speed at a given time.
+    :args: a list of (time, distance, speed) tuples.
```
Comment (danieloconell) [inline utilities/profile_generator.py]: Google style plz

File: utilities/profile_generator.py [hunk 26]
```diff
@@ -0,0 +1,217 @@
...
+                v = c[1]+2*c[2]*t_rel+3*c[3]*t_rel**2
+                a = 2*c[2]+6*c[3]*t_rel
+                return (d, v, a)
+    return trajectory
+
+
+def generate_interpolation_trajectory(x_start, x_final, traj_to_match):
+    """Generate a 1d interpolation profile, where the velocity is constant
+    over the duration of the trajectory.
+    :returns: a list of (pos, vel acc) tuples.
```
Comment (danieloconell) [inline utilities/profile_generator.py]: Google style

File: utilities/profile_generator.py [hunk 27]
```diff
@@ -0,0 +1,217 @@
...
+
+    def get_profile_point(distance, speed):
+        distance_proportion = distance / interpolation_distance
+        # convert from units/m to units / sec
+        vel = avg_vel * speed
+        return (x_start+distance_proportion*x, vel, 0)
+    return get_profile_point
+
+
+def generate_trapezoidal_function(
```
Comment (danieloconell) [inline utilities/profile_generator.py]: Docstring

File: utilities/profile_generator.py [hunk 28]
```diff
@@ -0,0 +1,217 @@
...
+    # how long we are cruising at v_max for (flat part of the trapezoid)
+    t_constant = t_decel - t_cruise
+    # how far we have travelled since the start when we start decelerating
+    x_decel = x_cruise + v_max*t_constant
+    # time at which we finish the trajetory
+
+    decel_dist = x_final - x_decel
+    decel_mag = v_max - v_final
+
+    def get_speed(distance):
```
Comment (danieloconell) [inline utilities/profile_generator.py]: Docstring

File: utilities/profile_generator.py [hunk 29]
```diff
@@ -0,0 +1,217 @@
...
+            print("WARNING: calling speed profile function when after final distance")
+            return v_final
+
+    return get_speed
+
+
+def generate_trapezoidal_trajectory(
+        x_start, v_start, x_final, v_final, v_max, a_pos, a_neg, frequency):
+    """Generate a 1d trapezoidal profile.
+    :returns: a list of (pos, vel acc) tuples
```
Comment (danieloconell) [inline utilities/profile_generator.py]: Google style

File: robot.py (hunk 20 above)
Comment (auscompgeek) [reply robot.py]: These comments are not documentation.

File: tests/test_vector_pursuit.py (hunk 21 above)
Comment (auscompgeek) [reply tests/test_vector_pursuit.py]: Do we really need docstrings in the tests?

File: autonomous/autonomous.py (hunk 5 above)
Comment (auscompgeek) [reply autonomous/autonomous.py]: I feel the intrinsic documentation (oh god SDD has gotten to me) here is enough.

File: pyswervedrive/swervechassis.py (hunk 18 above)
Comment (auscompgeek) [reply pyswervedrive/swervechassis.py]: It's… implementing the PIDOutput interface? What more do you want to know?

File: autonomous/autonomous.py (hunk 5 above)
Comment (ArthurAllshire) [reply autonomous/autonomous.py]: 👍 I think this applies to all of the autonomous stuff.

File: automations/motion.py (hunk 4 above)
Comment (ArthurAllshire) [reply automations/motion.py]: Why do you want a docstring on a property? Esp. one that is literally only used internally in the object

File: autonomous/autonomous.py [hunk 30]
```diff
@@ -0,0 +1,295 @@
...
+            self.next_state("go_to_scale")
+        elif self.been_to_switch is False:  # switch needs a cube and pickup was successful
+            self.next_state("go_to_switch")
+        if 1 == 1:  # After completing the intake cycle there is no cube
+            self.next_state("search_for_cube")
+
+
+class LeftSwitchAndScale(SwitchAndScale):
+    """The switch and scale strategy when we start on the left"""
+    MODE_NAME: 'Switch and scale - left start'
```
Comment (auscompgeek) [inline autonomous/autonomous.py]: This is broken.

File: autonomous/autonomous.py [hunk 31]
```diff
@@ -0,0 +1,295 @@
...
+    start_position = 3  # change to left start position
```
Comment (auscompgeek) [inline autonomous/autonomous.py]: What is this magic number?

File: automations/motion.py (hunk 3 above)
Comment (ArthurAllshire) [reply automations/motion.py]: Why do you want documentation on a function that is never intended to be called from outside this object?

File: autonomous/autonomous.py [hunk 32]
```diff
@@ -0,0 +1,295 @@
...
+            pass
+        elif self.fms_scale == 'L':
+            # go to left scale
+            pass
+        self.next_state("deposit_cube")
+
+
+class RightSwitchAndScale(SwitchAndScale):
+    """The switch and scale strategy when we start on the right"""
+    MODE_NAME: 'Switch and scale - right start'
```
Comment (auscompgeek) [inline autonomous/autonomous.py]: This is broken too.

File: pyswervedrive/swervechassis.py (hunk 2 above)
Comment (ArthurAllshire) [reply pyswervedrive/swervechassis.py]: This is a print statement...

File: autonomous/autonomous.py [hunk 33]
```diff
@@ -0,0 +1,295 @@
...
+                    [7.07, 1.90, 0, 1]],  # SameSwitchCrossScale
+                   [[7.5, -1.87, 0, 1],
+                    [5.5, -2, 2, 1],
+                    [6, 2.1, 4.71239, 1],
+                    [4.2, 2.1, 4.71239, 1]],  # CrossSwitchSameScale
+                   [[5, -3, 0, 1],
+                    [6, 2, 0, 1],
+                    [7.5, 1.8, 0, 1],
+                    [5.5, 2, 3.92699, 1],
+                    [5.25, 1.5, 3.14159, 1]]]  # CrossSwitchCrossScale
```
Comment (auscompgeek) [inline autonomous/autonomous.py]: Uh. What. Why are these here?

File: components/vision.py (hunk 11 above)
Comment (ArthurAllshire) [reply components/vision.py]: This function is obviously never intended to be called from outside vision. Why does it need to be documented?

File: autonomous/autonomous.py [hunk 34]
```diff
@@ -0,0 +1,295 @@
...
+            pass
+        elif self.fms_scale == 'L':
+            # go to left scale
+            pass
+        self.next_state("deposit_cube")
+
+
+class LeftDoubleScale(OverallBase):
+    """The double switch strategy when we start on the left"""
+    MODE_NAME: 'Double scale - left start'
```
Comment (auscompgeek) [inline autonomous/autonomous.py]: This is also broken.

File: autonomous/autonomous.py [hunk 35]
```diff
@@ -0,0 +1,295 @@
...
+        elif self.fms_scale == 'L':
+            self.coordinates = self.coordinates[0]
+            self.motion.set_waypoints = np.array(self.coordinates[0])
+            # go to left scale
+        self.next_state("deposit_cube")
+
+
+class RightDoubleScale(OverallBase):
+    """The double switch strategy when we start on the right"""
+    MODE_NAME: 'Double scale - right start'
```
Comment (auscompgeek) [inline autonomous/autonomous.py]: This too.

File: components/vision.py (hunk 12 above)
Comment (ArthurAllshire) [reply components/vision.py]: This function is obviously never intended to be called from outside this class. So why does it need to be documented?

File: components/vision.py [hunk 36]
```diff
@@ -0,0 +1,56 @@
...
+import numpy as np
+from networktables import NetworkTables
+
+
+class Vision:
+
+    def setup(self):
+        """This is called after variables are injected by magicbot."""
+        self.nt = NetworkTables
+        self.sd = self.nt.getTable("/vision")
```
Comment (auscompgeek) [inline components/vision.py]: What on earth is going on here?

File: components/vision.py [hunk 37]
```diff
@@ -0,0 +1,56 @@
...
+        info = self.format_list(info, 3)
+        total_time = (time.time() - start_time) + float(info[-1])
+        info[-1] = total_time
+
+        self.vision = info
+
+    def largest_cube(self):
+        if len(self.vision) < 2:
+            return None
+        return self.vision[0][0]
```
Comment (auscompgeek) [inline components/vision.py]: This entire file seems like a rather absurd way of doing things. It's like someone tried to apply Software Engineering Patterns™ to a problem without actually considering the problem first.

File: robot.py [hunk 38]
```diff
@@ -3,23 +3,33 @@
...
     # Add magicbot components here using variable annotations.
     # Any components that directly actuate motors should be declared after
     # any higher-level components (automations) that depend on them.
 
+    ds: wpilib.DriverStation
```
Comment (auscompgeek) [inline robot.py]: Wat.

File: utilities/vector_pursuit.py [hunk 39]
```diff
@@ -0,0 +1,102 @@
...
+from utilities.profile_generator import generate_trapezoidal_function
+import numpy as np
+
+
+class VectorPursuit:
+
+    def set_waypoints(self, waypoints: np.array):
+        """Add way points to.
+        Args:
+            waypoint list of numpy arrays.
```
Comment (auscompgeek) [inline utilities/vector_pursuit.py]: These docstrings are difficult to read. I think that may be because they aren't written properly.

File: robot.py [hunk 40]
```diff
@@ -57,18 +67,37 @@ def createObjects(self):
         # boilerplate setup for the joystick
         self.joystick = wpilib.Joystick(0)
 
+        self.cube_switch = wpilib.DigitalInput(0)
+
         self.spin_rate = 5
 
+        NetworkTables.initialize()
+        self.sd = NetworkTables.getTable("SmartDashboard")
```
Comment (auscompgeek) [inline robot.py]: Is this actually being used anywhere? (I think the answer is no, and I don't see a very good reason for doing this.)

File: pyswervedrive/swervechassis.py (hunk 18 above)
Comment (auscompgeek) [inline pyswervedrive/swervechassis.py]: I think it might actually be a good idea to document what ChassisPIDOutput is being used for. Looking at this without context, one might not actually realise what's going on here.

File: autonomous/autonomous.py [hunk 41]
```diff
@@ -0,0 +1,295 @@
...
+    """A less general routine for the switch and scale strategy. Still requires subclassing"""
+
+    def __init__(self):
+        self.been_to_switch = False
+        super().__init__()
+
+    @state
+    def go_to_switch(self):
+        """Goes to the switch, when subclassed will go to the correct side regardless of start."""
+        self.been_to_switch: True
```
Comment (auscompgeek) [inline autonomous/autonomous.py]: wat.

File: autonomous/autonomous.py [hunk 42]
```diff
@@ -0,0 +1,295 @@
...
+            pass
+        self.next_state("deposit_cube")
+
+    @state
+    def intake_cube(self):
+        """Attempts to intake the cube"""
+        # Run intake
+        if self.been_to_switch:  # cube inside intake system (current spikes)
+            self.next_state("go_to_scale")
+        elif self.been_to_switch is False:  # switch needs a cube and pickup was successful
```
Comment (auscompgeek) [inline autonomous/autonomous.py]: `else:`

File: autonomous/autonomous.py [hunk 43]
```diff
@@ -0,0 +1,295 @@
...
+    def go_to_scale(self):
+        if self.fms_scale == 'R':
+            self.coordinates = self.coordinates[0]
+            self.motion.set_waypoints = np.array(self.coordinates[0])
+            # go to right scale
+        elif self.fms_scale == 'L':
+            self.coordinates = self.coordinates[1]
+            self.motion.set_waypoints = np.array(self.coordinates[0])
+            self.motion.set_waypoints = np.array(self.coordinates[1])
+            self.motion.set_waypoints = np.array(self.coordinates[2])
```
Comment (auscompgeek) [inline autonomous/autonomous.py]: ???

File: pyswervedrive/swervemodule.py [hunk 44]
```diff
@@ -1,32 +1,32 @@
...
     CIMCODER_COUNTS_PER_REV: int = 80
     WHEEL_DIAMETER: float = 0.0254 * 3
-    DRIVE_ENCODER_GEAR_REDUCTION: float = 5.43956
+    DRIVE_ENCODER_GEAR_REDUCTION: float = 66/14*30/26
     STEER_COUNTS_PER_RADIAN = 4096 / math.tau
 
     drive_counts_per_rev = CIMCODER_COUNTS_PER_REV*DRIVE_ENCODER_GEAR_REDUCTION
     drive_counts_per_radian = drive_counts_per_rev / math.tau
-    drive_counts_per_metre = drive_counts_per_rev / (math.pi * WHEEL_DIAMETER)
+    drive_counts_per_metre = drive_counts_per_rev / (math.pi * WHEEL_DIAMETER) / 1.08
```
Comment (auscompgeek) [inline pyswervedrive/swervemodule.py]: What's this magic number at the end?

File: pyswervedrive/swervemodule.py [hunk 45]
```diff
@@ -39,38 +39,39 @@ def __init__(self, steer_talon: ctre.WPI_TalonSRX, drive_talon: ctre.WPI_TalonSR
...
         # changes sign of motor throttle vilues
         self.steer_motor.setInverted(self.reverse_steer_direction)
         sp = self.steer_motor.getSelectedSensorPosition(0)
         self.current_azimuth_sp = float(sp - self.steer_enc_offset) / self.STEER_COUNTS_PER_RADIAN
-        self.steer_motor.config_kP(0, 0.0001, 10)
+        self.steer_motor.config_kP(0, 0.1, 10)
         self.steer_motor.config_kI(0, 0.0, 10)
         self.steer_motor.config_kD(0, 0.0, 10)
         self.steer_motor.selectProfileSlot(0, 0)
         self.steer_motor.config_kF(0, 0, 10)
```
Comment (auscompgeek) [inline pyswervedrive/swervemodule.py]: Looking at all these config calls again, these really should make it more obvious the last parameter is the timeout.

File: pyswervedrive/swervemodule.py [hunk 46]
```diff
@@ -166,24 +167,29 @@ def set_velocity(self, vx, vy):
...
             # figure out the most efficient way to get the module to the desired direction
-            current_unwound_azimuth = constrain_angle(self.current_measured_azimuth)
+            # current_unwound_azimuth = constrain_angle(self.current_measured_azimuth)
+            current_unwound_azimuth = constrain_angle(self.current_azimuth_sp)
             delta = self.min_angular_displacement(current_unwound_azimuth, desired_azimuth)
 
-        # # Please note, this is *NOT WRAPPED* to +-pi, because if wrapped the module
-        # # will unwind
+        # # # Please note, this is *NOT WRAPPED* to +-pi, because if wrapped the module
+        # # # will unwind
```
Comment (auscompgeek) [inline pyswervedrive/swervemodule.py]: Do you have enough `# ` at the beginning of these lines yet?

File: autonomous/autonomous.py (hunk 30 above)
Comment (ArthurAllshire) [reply autonomous/autonomous.py]: @mlists can you fix the review changes for autonomous in your repo, and I will put the fixes into my PR

File: components/vision.py (hunk 37 above)
Comment (ArthurAllshire) [reply components/vision.py]: @OliverAllshire @Ninjakow  can you talk to David about how to fix the vision component?

File: pyswervedrive/swervechassis.py (hunk 18 above)
Comment (ArthurAllshire) [reply pyswervedrive/swervechassis.py]: Well, if you look at the rest of the file, I don't actually ever use the output given to this object (I use `controller.get()` instead), because I think that this way is unclear anyway. But maybe should document that it isnt actually being used for anything.

File: utilities/vector_pursuit.py (hunk 39 above)
Comment (ArthurAllshire) [reply utilities/vector_pursuit.py]: Oops, I haven't updated this docstring since Daniel's firsts pass at this code. Will update it.

File: pyswervedrive/swervemodule.py (hunk 44 above)
Comment (ArthurAllshire) [reply pyswervedrive/swervemodule.py]: Fudge factor for the odometry. Will comment that.

File: autonomous/autonomous.py (hunk 5 above)
Comment (danieloconell) [reply autonomous/autonomous.py]: Agreed

File: utilities/profile_generator.py [hunk 47]
```diff
@@ -0,0 +1,225 @@
...
+    else:
+        return -1
+
+
+def cubic_generator(keypoints):
+    """Return a function that returns the distance and speed at a given time.
+    Args:
+        keypoints: a list of (time, distance, speed) tuples.
+    """
+
```
Comment (auscompgeek) [inline utilities/profile_generator.py]: no blank line after function docstring
